[E-Documents Core] - Migrate electronic formats to new purchase draft concept for e-invoicing (e-invoice preview) (PEPPOL)

GMatuleviciute opened the pull request:
This pull request does not have a related issue as it's part of the delivery for development agreed directly with @altotovi @Groenbech96

**Implementation**

- Introduced new enum and interface for handling Purchase Credit Memos.
- Updated Purchase Invoice codeunit to validate and handle additional fields.
- Enhanced E-Document Purchase Header table to include E-Document Type.
- Implemented logic in PEPPOL handler for processing Credit Notes.
- Added methods for populating Credit Memo headers and lines in XML.
- Refactored existing methods to utilize a new XML helper codeunit for cleaner code. (This codeunit enables localization apps to reuse same procedures as used in W1)
- Updated unit of measure provider interface to include retrieval of default unit of measure.

Fixes #

Fixes [AB#598453](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/598453)

AndriusAndrulevicius committed:
feat: Add support for Purchase Credit Memo in E-Document processing

- Introduced new enum and interface for handling Purchase Credit Memos.
- Updated Purchase Invoice codeunit to validate and handle additional fields.
- Enhanced E-Document Purchase Header table to include E-Document Type.
- Implemented logic in PEPPOL handler for processing Credit Notes.
- Added methods for populating Credit Memo headers and lines in XML.
- Refactored existing methods to utilize a new XML helper codeunit for cleaner code.
- Updated unit of measure provider interface to include retrieval of default unit of measure.

AndriusAndrulevicius committed:
refactor: Simplify Purchase Credit Memo creation and enhance XML processing

AndriusAndrulevicius committed:
Fixes purchase line numbering by using FindLast

Changes the purchase line lookup from FindFirst to FindLast to ensure proper line number sequencing when creating new purchase lines.

Moves the CreatePurchaseLine method to the end of the file for better code organization.

AndriusAndrulevicius committed:
Adds XML documentation to E-Document helper methods

Improves code maintainability by adding comprehensive XML documentation comments to helper methods in the E-Document module.

Documents parameter types, descriptions, and return values for XML parsing utilities and purchase credit memo creation procedures to enhance developer experience and code understanding.

AndriusAndrulevicius committed:
Update EDocumentXMLHelper.Codeunit.al

Use next available ID (recent MSFT changes introduced new obects)

AndriusAndrulevicius committed:
merge main into dev/aan/PEPPOL-v2

AndriusAndrulevicius committed:
Reorganizes imports and updates codeunit ID

Changes codeunit number from 6175 to 6177 to resolve potential ID conflicts.

Alphabetizes using statements across multiple files to improve code organization and maintainability.

Adds missing blank lines in enum definitions for better formatting consistency.

Enhances documentation by adding parameter descriptions to interface method.

AndriusAndrulevicius committed:
Adds whitespace for code readability

Improves code formatting by adding a blank line after the exit statement to enhance visual separation between the early return and subsequent logic flow.

AndriusAndrulevicius committed:
Improves unit of measure handling in purchase documents

Sets default unit of measure when missing during purchase line processing to prevent validation errors.

Clears item reference number when purchase type changes to maintain data consistency and avoid mismatched references.

GMatuleviciute committed:
Apply suggestion from @GMatuleviciute

GMatuleviciute committed:
Apply suggestion from @GMatuleviciute

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/FinishDraft/EDocCreatePurchaseInvoice.Codeunit.al:
```diff
@@ -106,6 +105,7 @@ codeunit 6117 "E-Doc. Create Purchase Invoice" implements IEDocumentFinishDraft,
         end;
 
         PurchaseHeader.Validate("Vendor Invoice No.", VendorInvoiceNo);
+        PurchaseHeader.Validate("Vendor Order No.", EDocumentPurchaseHeader."Purchase Order No.");
```
Vendor Order No. refers to external number for seller. Purchase Order No. on EDocumentPurchaseHeader refers to internal numbering system.

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/StructureReceivedEDocument/EDocumentPEPPOLHandler.Codeunit.al:
```diff
@@ -81,126 +85,152 @@ codeunit 6173 "E-Document PEPPOL Handler" implements IStructuredFormatReader
-            exit;
-        end;
+        EDocumentXMLHelper.SetStringValueInField(PeppolXML, XMLNamespaces, '/cn:CreditNote/cbc:ID', MaxStrLen(EDocumentPurchaseHeader."Sales Invoice No."), EDocumentPurchaseHeader."Sales Invoice No.");
+        EDocumentXMLHelper.SetStringValueInField(PeppolXML, XMLNamespaces, '/cn:CreditNote/cac:BillingReference/cac:InvoiceDocumentReference/cbc:ID', MaxStrLen(EDocumentPurchaseHeader."Purchase Order No."), EDocumentPurchaseHeader."Purchase Order No.");
```
This is not correct. 
This refers to a invoice in our system, that the credit memo applied to. 
We actually dont have such a field on the edoc purch header, but we should have so we can read this number into it.

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/StructureReceivedEDocument/EDocumentPEPPOLHandler.Codeunit.al:
```diff
@@ -81,126 +85,152 @@ codeunit 6173 "E-Document PEPPOL Handler" implements IStructuredFormatReader
-#pragma warning restore AA0139
 
-    local procedure SetStringValueInField(XMLDocument: XmlDocument; XMLNamespaces: XmlNamespaceManager; Path: Text; MaxLength: Integer; var Field: Text)
+    local procedure PopulatePurchaseCreditMemoHeader(PeppolXML: XmlDocument; XmlNamespaces: XmlNamespaceManager; var EDocumentPurchaseHeader: Record "E-Document Purchase Header")
```
Any reason why CreditNote/cac:OrderReference/cbc:ID is not parsed?

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/FinishDraft/EDocCreatePurchCrMemo.Enum.al:
```diff
@@ -0,0 +1,18 @@
+/// <summary>
+///  Enum for the implementations of the E-Doc. Create Purchase Credit Memo interface.
+/// </summary>
+enum 6118 "E-Doc. Create Purch. Cr. Memo" implements IEDocumentCreatePurchaseCreditMemo
```
Can we put the two create enums together into Create BC Docment, and then have 1 entry for invoice, and 1 entry for credit note? 

That would mean to obsolete the E-Doc. Create Purch. Invoice enum, which is fine. 
Otherwise for each document type we would have a new enum, which is not really needed.

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/FinishDraft/EDocCreatePurchaseInvoice.Enum.al:
```diff
@@ -13,7 +13,6 @@ enum 6105 "E-Doc. Create Purchase Invoice" implements IEDocumentCreatePurchaseIn
-    {
-    }
+    
+    value(0; "Default") { }
```
See comment above about obsolete.

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/ImportEDocumentProcess.Codeunit.al:
```diff
@@ -184,6 +185,11 @@ codeunit 6104 "Import E-Document Process"
                 end;
+            Step::"Read into Draft":
+                begin
+                    EDocumentPurchaseLine.SetRange("E-Document Entry No.", EDocument."Entry No");
```
We will need to keep **purchase specific** stuff out of this.

Can we add undo function to the interface for reading into draft.

AndriusAndrulevicius replied:
fixed

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/PrepareDraft/EDocProviders.Codeunit.al:
```diff
@@ -132,8 +133,20 @@ codeunit 6124 "E-Doc. Providers" implements IPurchaseLineProvider, IUnitOfMeasur
-        if PurchaseHeader.Get("Purchase Document Type"::Order, EDocumentPurchaseHeader."Purchase Order No.") then;
+        Vendor.SetLoadFields("Receive E-Document To");
+        Vendor.Get(EDocumentPurchaseHeader."[BC] Vendor No.");
+        case Vendor."Receive E-Document To" of
```
We should exit if Receive to invoice. This function should never return a purchase order. :) 
I think we can keep it at that for now.

AndriusAndrulevicius replied:
the original code was retrieving Order - I've just implemented the logic to retrieve the document type based on settings on Vendor card.

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/PrepareDraft/EDocProviders.Codeunit.al:
```diff
@@ -171,4 +184,25 @@ codeunit 6124 "E-Doc. Providers" implements IPurchaseLineProvider, IUnitOfMeasur
+        Item: Record Item;
+        Resource: Record Resource;
+    begin
+        case PurchaseLineType of
```
Can we move this up to when we find an item reference without UOM? 
That would mean move this code up somewhere at line 184 and change that a bit.

Groenbech96 reviewed (changes requested):
Really good work on this. We have reviewed and left some feedback that should be addressed.

AndriusAndrulevicius committed:
Enhances PEPPOL e-document testing with comprehensive validation

Adds three new test methods to validate PEPPOL invoice processing workflow including data extraction viewing, purchase invoice creation, and draft modification scenarios.

Improves test data consistency by introducing mock currency and date variables that can be configured across test methods rather than hardcoded values.

Fixes page validation trigger formatting by removing unnecessary page update call and correcting indentation.

Strengthens assertion messages with detailed error information including field names, table names, expected and actual values for better debugging.

AndriusAndrulevicius committed:
Refactors purchase document creation logic

Consolidates duplicate code between invoice and credit memo creation into a shared helper class to improve maintainability and consistency.

Adds ResetDraft method to structured format readers for proper cleanup when undoing processing steps.

Removes obsolete enum and simplifies unit of measure handling by moving default logic to item reference processing.

Enhances credit memo creation with applies-to document support for better document linking.

AndriusAndrulevicius commented on Apps/W1/EDocument/app/src/Processing/Import/FinishDraft/EDocCreatePurchCrMemo.Codeunit.al:
```diff
@@ -0,0 +1,224 @@
+        EDocumentPurchaseLine.SetRange("E-Document Entry No.", EDocument."Entry No");
+        if EDocumentPurchaseLine.FindSet() then
+            repeat
+                CreatePurchaseLine(EDocument, PurchaseHeader, EDocumentPurchaseLine, EDocumentPurchaseHistMapping);
```
fixed

AndriusAndrulevicius committed:
Adds XML node value retrieval helper method

Introduces a utility method that simplifies extracting text content from XML nodes using XPath expressions.

The method handles node selection and safely returns the inner text value or an empty string if the node is not found, reducing boilerplate code for XML parsing operations.

AndriusAndrulevicius committed:
Refactors draft reset logic into reusable method

Consolidates duplicate draft reset code by moving the implementation to a central ResetDraft method on the E-Document table. Multiple codeunits previously contained identical logic for deleting purchase headers, lines, mappings, and line fields.

Improves maintainability by eliminating code duplication and providing a single source of truth for draft cleanup operations.

AndriusAndrulevicius committed:
Fixes field mapping in PEPPOL credit note XML generation

Corrects the assignment of purchase order number and applies-to document number fields in credit note XML structure.

Previously, the values were swapped between OrderReference and BillingReference sections, causing incorrect document references in generated PEPPOL credit notes.

AndriusAndrulevicius committed:
Prevents errors when deleting empty record sets

Adds empty checks before calling delete operations to avoid potential runtime errors when attempting to delete records that don't exist.

Improves robustness by ensuring delete operations are only performed when there are actual records to delete.

AndriusAndrulevicius committed:
Update Apps/W1/EDocument/test/src/Processing/PEPPOLStructuredValidations.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/W1/EDocument/test/src/Processing/EDocumentStructuredTests.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/W1/EDocument/test/src/Processing/EDocumentStructuredTests.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/W1/EDocument/test/src/Processing/EDocumentStructuredTests.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/W1/EDocument/test/src/Processing/EDocumentStructuredTests.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/W1/EDocument/test/src/Processing/EDocumentStructuredTests.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/W1/EDocument/test/src/Processing/EDocumentStructuredTests.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

AndriusAndrulevicius committed:
Update Apps/W1/EDocument/test/src/Processing/EDocumentStructuredTests.Codeunit.al

Co-authored-by: Grasiele Matuleviciute <131970463+GMatuleviciute@users.noreply.github.com>

Groenbech96 commented on Apps/W1/EDocument/app/src/Document/EDocument.Table.al:
```diff
@@ -479,6 +479,36 @@ table 6121 "E-Document"
+
+        EDocumentLineField.SetRange("E-Document Entry No.", Rec."Entry No");
+        if not EDocumentLineField.IsEmpty() then
+            EDocumentLineField.DeleteAll(true);
```
DeleteAll does not need guard.

AndriusAndrulevicius committed:
Makes PEPPOL validations stateless and consistent

Refactors validation helpers to avoid shared state by passing required data explicitly, reducing test coupling and flakiness.

Unifies and locks label tokens, standardizes error messages, and fixes API usage for captions. Cleans up unused variables and minor formatting. Updates tests to use the new helper signature.

AndriusAndrulevicius committed:
Merge branch 'main' into dev/aan/PEPPOL-v2

AndriusAndrulevicius committed:
Centralizes draft reset on purchase header

Moves draft cleanup from the document level to the purchase header, deleting header, lines, header mappings, and additional line fields in one place.

Updates import handlers to fetch the purchase header from the document and invoke the new reset, improving encapsulation and avoiding partial cleanup.

Aligns a history-mapping call with the updated API naming for clarity.

AndriusAndrulevicius committed:
Enhances PEPPOL validations with stateless and consistent testing framework

Implements comprehensive PEPPOL e-document validation improvements including:
- Stateless and consistent PEPPOL validation implementation
- Enhanced structured e-document testing framework
- Improved test coverage for PEPPOL validation scenarios
- Refined validation logic for better consistency across operations

AndriusAndrulevicius committed:
Centralizes draft cleanup; adds invoice retrieval

Moves draft cleanup to the record delete trigger to always remove related lines and mappings when discarding a draft, removing the custom reset method and updating callers to delete the header directly.

Splits purchase document lookup by exposing separate methods for order and invoice. The caller now chooses based on vendor setup, decoupling provider logic from vendor configuration and avoiding provider-level validation errors.

Minor refactor in draft preparation to handle the new lookup and streamline flow.

Groenbech96 commented on Apps/W1/EDocument/app/src/Processing/Import/PrepareDraft/PreparePurchaseEDocDraft.Codeunit.al:
```diff
@@ -48,11 +48,16 @@ codeunit 6125 "Prepare Purchase E-Doc. Draft" implements IProcessStructuredData
         end;
 
-        PurchaseOrder := IPurchaseOrderProvider.GetPurchaseOrder(EDocumentPurchaseHeader);
+        case Vendor."Receive E-Document To" of
```
I would use this in the exit clause on line 62. 
exit(if Vendor.Receive E-Document To = Purchase Order ? E-Document Type::PO : E-Document Type::Invoice)

Then the rest can be left as it was. This setting should probably only influence that bit. Even if we have invoice, it would be okay to fill [BC] Purchase Order No.

AndriusAndrulevicius committed:
Merge branch 'main' into dev/aan/PEPPOL-v2

AndriusAndrulevicius committed:
Respects vendor target: order vs invoice in drafts

Updates draft preparation to honor the vendor’s target (order or invoice), validating document type and returning the corresponding e-document type instead of always using order.

Ensures purchase header data is initialized from the e-document before processing to prevent missing header values.

Adapts tests to the parameter-based import call and updates helpers/mocks accordingly. Adds attachment dependency for purchase invoice creation.

AndriusAndrulevicius reviewed (commented):
All comments should be now resolved.
Automated tests also included in this PR

JesperSchulz commented:
This has been merged internally. Closing PR.
